lib/Dialect/D2M/Utils/Utils.cpp

```diff
@@ -128,4 +132,48 @@ AffineMap concatInversePermutationMap(mlir::ArrayRef<AffineMap> affineMaps,
   return mlir::inversePermutation(concat);
 }
 
+Value getPhysicalTensorOrMemref(mlir::Value tensorOrMemref) {
+
+  TT_assertv((mlir::isa<mlir::RankedTensorType>(tensorOrMemref.getType()) ||
+              mlir::isa<mlir::MemRefType>(tensorOrMemref.getType())),
+             "Expected a tensor or memref type");
+  auto physTensor = tensorOrMemref;
+
+  if (auto viewOp = mlir::dyn_cast<mlir::tt::d2m::ViewOpInterface>(
+          tensorOrMemref.getDefiningOp())) {
+    physTensor = viewOp.getInput();
+  } else if (auto toLayoutOp = mlir::dyn_cast<mlir::tt::d2m::ToLayoutOp>(
+                 tensorOrMemref.getDefiningOp())) {
+    return getPhysicalTensorOrMemref(toLayoutOp.getInitOperand());
+
+  } else if (auto genericOp = mlir::dyn_cast<mlir::tt::d2m::GenericOp>(
+                 tensorOrMemref.getDefiningOp())) {
+    // Assume that if the defining op is a generic op, the output is the first
+    // of the outputs.
+    physTensor = getPhysicalTensorOrMemref(genericOp.getOutputs()[0]);
+  }
+
+  return physTensor;
+}
+
+ArrayRef<int64_t> getGridShape(Value tensorOrMemref) {
+  TT_assertv((mlir::isa<RankedTensorType>(tensorOrMemref.getType()) ||
+              mlir::isa<MemRefType>(tensorOrMemref.getType())),
+             "Expected a tensor or memref type");
+
+  Attribute layout =
+      (mlir::isa<RankedTensorType>(tensorOrMemref.getType())
+           ? mlir::dyn_cast<RankedTensorType>(tensorOrMemref.getType())
+                 .getEncoding()
+           : mlir::dyn_cast<MemRefType>(tensorOrMemref.getType()).getLayout());
+
+  if (auto deviceLayout =
+          mlir::dyn_cast_if_present<ttcore::DeviceLayoutInterface>(layout)) {
+    return deviceLayout.getGridShape(
+        mlir::dyn_cast<ShapedType>(tensorOrMemref.getType()));
+  } else {
+    return {};
+  }
```

Comment:
:warning: **llvm\-else\-after\-return** :warning:
do not use `` else `` after `` return ``
```suggestion
  }     return {};
 
```

Reply:
Fixed.